File: src/net/auoeke/cin/lexer/Context.java
package net.auoeke.cin.lexer;

public enum Context {
    FILE,
    ARRAY,
    MAP;

    public char start() {
        return switch (this) {
            case ARRAY -> '[';
            case MAP -> '{';
            case FILE -> Character.MIN_VALUE;
        };
    }

    public String description() {
        return switch (this) {
            case ARRAY -> "an array";
            case MAP -> "a map";
            case FILE -> "a file";
        };
    }

    public static Context of(char character) {
        return switch (character) {
            case '[' -> ARRAY;
            case '{' -> MAP;
            default -> throw new Error();
        };
    }
}

File: src/net/auoeke/cin/lexer/Context.kt
package net.auoeke.cin.lexer

enum class Context {
    FILE,
    ARRAY,
    MAP;

    val start: Char get() = when (this) {
        ARRAY -> '['
        MAP -> '{'
        FILE -> Char.MIN_VALUE
    }

    val description: String get() = when (this) {
            ARRAY -> "an array"
            MAP -> "a map"
            FILE -> "a file"
        }

    companion object {
        val Char.context: Context get() = when (this) {
            '[' -> ARRAY
            '{' -> MAP
            else -> throw Error()
        }
    }
}
